### DISCUSSION TITLE
Flux and dynamically scaling in and out

### DISCUSSION THREAD

**Discussion by stevendick-work** [fluxcd/flux2#4007] [Q&A] [answered]
My project is starting to look at dynamically scale our cluster in and out to match the workload. I am wondering how this can work given that:

- Our Flux HelmRelease resources specify the number of instances for the deployment
- An autoscaling solution like KEDA might have changed the number of instances for the deployment since the last reconciliation
- A new Flux reconciliation will reset the instances to match the HelmRelease

Has anyone implemented something like KEDA in a cluster that is also using Flux?

**Comment by kingdonb** [comment, marked as answer]
Hi there! I am a KEDA user and I've been experimenting with the HTTP Add-On and Flux with KEDA. My setup broke recently due to some updates in the http-addon, I'm still working that out. But I have an idea about how this works. Today at Bug Scrub, we had an experiment to try ChatGPT. We fed your question to GPT4 and were surprised by the apparent high quality of the answer, so I'm going to paste it here.

The essence of the answer appears to be absolutely correct IMO, but beware the code samples may not be any good as they were generated by a robot (Unfortunately it also messes up the formatting and the links to do this paste directly):

---

Hello,

It's great to see that you're exploring autoscaling solutions like KEDA alongside Flux v2. You can definitely use KEDA and Flux together to manage autoscaling in your Kubernetes cluster. To achieve this, you simply need to avoid specifying the number of replicas directly in the HelmRelease resource. Instead, you can configure autoscaling in your Helm chart using a HorizontalPodAutoscaler (HPA) or KEDA ScaledObject.

Here's a high-level overview of how to set this up:

Update your Helm chart: Modify your Helm chart to include a HorizontalPodAutoscaler or KEDA ScaledObject resource. Remove any hardcoded replica count from your Deployment in the chart, and use a value from the chart's `values.yaml` file to make it configurable.

Update your HelmRelease: Configure the HelmRelease resource in your Git repository with the desired default replica count and autoscaling settings. Flux will apply these settings when reconciling your cluster.

Autoscaling takes over: KEDA or the HPA will monitor the metrics specified in your autoscaling configuration, and scale your deployment in and out accordingly. Since you're no longer specifying the number of instances directly in the HelmRelease, Flux won't reset the replica count during reconciliation.

Here's a simplified example of a KEDA ScaledObject that you can use as a starting point:

```yaml
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: my-scaled-object
spec:
  scaleTargetRef:
    name: my-deployment
  minReplicaCount: {{ .Values.minReplicaCount }}
  maxReplicaCount: {{ .Values.maxReplicaCount }}
  triggers:
  - type: my-trigger-type
    metadata:
      # Trigger-specific settings
```

Remember to replace my-deployment with the name of your deployment, and update the trigger settings as needed.

By following these steps, you should be able to use KEDA or HPA for autoscaling while still benefiting from Flux's GitOps capabilities.

Good luck with your project, and let me know if you have any questions!

---

That was all from the ChatGPT4 response. Let us know if this helps at all, and if you have any questions for a human, I'll do my best to get KEDA example working on one of my clusters again so I can test and check any details.

**Reply by stevendick-work** [reply]
For a v1, we are going for a more static solution of applying changes to the specified number of instances in the Git repo, but it's good to know there is a pattern for working with KEDA when we get there.
